// @base-ui/react/otp-field/root/OTPFieldRootDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OTPFieldRootDataAttributes = void 0;
let OTPFieldRootDataAttributes = exports.OTPFieldRootDataAttributes = /*#__PURE__*/function (OTPFieldRootDataAttributes) {
  /**
   * Present when all slots are filled.
   */
  OTPFieldRootDataAttributes["complete"] = "data-complete";
  /**
   * Present when the OTP field is disabled.
   */
  OTPFieldRootDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the OTP field is readonly.
   */
  OTPFieldRootDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the OTP field is required.
   */
  OTPFieldRootDataAttributes["required"] = "data-required";
  /**
   * Present when the OTP field is in a valid state (when wrapped in Field.Root).
   */
  OTPFieldRootDataAttributes["valid"] = "data-valid";
  /**
   * Present when the OTP field is in an invalid state (when wrapped in Field.Root).
   */
  OTPFieldRootDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the OTP field has been touched (when wrapped in Field.Root).
   */
  OTPFieldRootDataAttributes["touched"] = "data-touched";
  /**
   * Present when the OTP field's value has changed (when wrapped in Field.Root).
   */
  OTPFieldRootDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the OTP field contains at least one character.
   */
  OTPFieldRootDataAttributes["filled"] = "data-filled";
  /**
   * Present when one of the OTP field inputs is focused.
   */
  OTPFieldRootDataAttributes["focused"] = "data-focused";
  return OTPFieldRootDataAttributes;
}({});
