// @base-ui/react/autocomplete/input-group/AutocompleteInputGroupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AutocompleteInputGroupDataAttributes = void 0;
let AutocompleteInputGroupDataAttributes = exports.AutocompleteInputGroupDataAttributes = /*#__PURE__*/function (AutocompleteInputGroupDataAttributes) {
  /**
   * Present when the corresponding popup is open.
   */
  AutocompleteInputGroupDataAttributes["popupOpen"] = "data-popup-open";
  /**
   * Present when the input group is pressed.
   */
  AutocompleteInputGroupDataAttributes["pressed"] = "data-pressed";
  /**
   * Present when the component is disabled.
   */
  AutocompleteInputGroupDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the component is readonly.
   */
  AutocompleteInputGroupDataAttributes["readonly"] = "data-readonly";
  /**
   * Indicates which side the corresponding popup is positioned relative to its anchor.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start' | null}
   */
  AutocompleteInputGroupDataAttributes["popupSide"] = "data-popup-side";
  /**
   * Present when the component is in a valid state (when wrapped in Field.Root).
   */
  AutocompleteInputGroupDataAttributes["valid"] = "data-valid";
  /**
   * Present when the component is in an invalid state (when wrapped in Field.Root).
   */
  AutocompleteInputGroupDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the component has been touched (when wrapped in Field.Root).
   */
  AutocompleteInputGroupDataAttributes["touched"] = "data-touched";
  /**
   * Present when the component's value has changed (when wrapped in Field.Root).
   */
  AutocompleteInputGroupDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the component has a value (when wrapped in Field.Root).
   */
  AutocompleteInputGroupDataAttributes["filled"] = "data-filled";
  /**
   * Present when the component is focused (when wrapped in Field.Root).
   */
  AutocompleteInputGroupDataAttributes["focused"] = "data-focused";
  /**
   * Present when the corresponding items list is empty.
   */
  AutocompleteInputGroupDataAttributes["listEmpty"] = "data-list-empty";
  return AutocompleteInputGroupDataAttributes;
}({});
