// @base-ui/react/esm/autocomplete/trigger/AutocompleteTriggerDataAttributes.js
import { CommonTriggerDataAttributes } from "../../utils/popupStateMapping.js";
export let AutocompleteTriggerDataAttributes = function (AutocompleteTriggerDataAttributes) {
  /**
   * Present when the corresponding popup is open.
   */
  AutocompleteTriggerDataAttributes[AutocompleteTriggerDataAttributes["popupOpen"] = CommonTriggerDataAttributes.popupOpen] = "popupOpen";
  /**
   * Present when the trigger is pressed.
   */
  AutocompleteTriggerDataAttributes[AutocompleteTriggerDataAttributes["pressed"] = CommonTriggerDataAttributes.pressed] = "pressed";
  /**
   * Present when the component is disabled.
   */
  AutocompleteTriggerDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the component is readonly.
   */
  AutocompleteTriggerDataAttributes["readonly"] = "data-readonly";
  /**
   * Indicates which side the corresponding popup is positioned relative to its anchor.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start' | null}
   */
  AutocompleteTriggerDataAttributes["popupSide"] = "data-popup-side";
  /**
   * Present when the component is required.
   */
  AutocompleteTriggerDataAttributes["required"] = "data-required";
  /**
   * Present when the component is in a valid state (when wrapped in Field.Root).
   */
  AutocompleteTriggerDataAttributes["valid"] = "data-valid";
  /**
   * Present when the component is in an invalid state (when wrapped in Field.Root).
   */
  AutocompleteTriggerDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the component has been touched (when wrapped in Field.Root).
   */
  AutocompleteTriggerDataAttributes["touched"] = "data-touched";
  /**
   * Present when the component's value has changed (when wrapped in Field.Root).
   */
  AutocompleteTriggerDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the component has a value (when wrapped in Field.Root).
   */
  AutocompleteTriggerDataAttributes["filled"] = "data-filled";
  /**
   * Present when the trigger is focused (when wrapped in Field.Root).
   */
  AutocompleteTriggerDataAttributes["focused"] = "data-focused";
  /**
   * Present when the corresponding items list is empty.
   */
  AutocompleteTriggerDataAttributes["listEmpty"] = "data-list-empty";
  return AutocompleteTriggerDataAttributes;
}({});
