// @base-ui/react/toast/root/ToastRootDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToastRootDataAttributes = void 0;
var _stateAttributesMapping = require("../../internals/stateAttributesMapping");
let ToastRootDataAttributes = exports.ToastRootDataAttributes = function (ToastRootDataAttributes) {
  /**
   * Present when the toast is expanded in the viewport.
   * @type {boolean}
   */
  ToastRootDataAttributes["expanded"] = "data-expanded";
  /**
   * Present when the toast was removed due to exceeding the limit.
   * @type {boolean}
   */
  ToastRootDataAttributes["limited"] = "data-limited";
  /**
   * The type of the toast.
   * @type {string}
   */
  ToastRootDataAttributes["type"] = "data-type";
  /**
   * Present when the toast is being swiped.
   * @type {boolean}
   */
  ToastRootDataAttributes["swiping"] = "data-swiping";
  /**
   * The direction the toast was swiped.
   * @type {'up' | 'down' | 'left' | 'right'}
   */
  ToastRootDataAttributes["swipeDirection"] = "data-swipe-direction";
  /**
   * Present when the toast is animating in.
   */
  ToastRootDataAttributes[ToastRootDataAttributes["startingStyle"] = _stateAttributesMapping.TransitionStatusDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the toast is animating out.
   */
  ToastRootDataAttributes[ToastRootDataAttributes["endingStyle"] = _stateAttributesMapping.TransitionStatusDataAttributes.endingStyle] = "endingStyle";
  return ToastRootDataAttributes;
}({});
