// @apps-in-toss/native-modules/src/MiniAppModule/native-modules/requestNotificationAgreement.ts
import { noop } from 'es-toolkit';
import { isMinVersionSupported } from './isMinVersionSupported';
import { INTERNAL__appBridgeHandler } from '../native-event-emitter/internal/appBridge';

const MIN_VERSION = { android: '5.255.0', ios: '5.255.0' } as const;

function isRequestNotificationAgreementSupported() {
  return isMinVersionSupported(MIN_VERSION);
}

/**
 * @public
 * @category 알림
 * @name NotificationAgreementResult
 * @description 푸시 알림 동의 플로우의 결과예요.
 * @property {'newAgreement'} newAgreement - 이번에 새로 동의한 경우예요.
 * @property {'alreadyAgreed'} alreadyAgreed - 이미 동의한 사용자인 경우예요.
 * @property {'agreementRejected'} agreementRejected - 사용자가 동의를 거부한 경우예요.
 */
export type NotificationAgreementResult = 'newAgreement' | 'alreadyAgreed' | 'agreementRejected';

/**
 * @property {string} templateCode - 알림 동의를 받을 스마트발송 캠패인의 템플릿 코드예요. 코드는 앱인토스 콘솔 > 미니앱 > 스마트발송에서 작성한 값을 입력해주세요.
 * @property {(result: { type: NotificationAgreementResult }) => void} onEvent - 알림 동의 결과를 콜백으로 전달해요.
 * @property {(error: unknown) => void | Promise<void>} onError - 알림 동의 요청 시 에러가 발생하면 콜백으로 전달해요.
 */
export interface RequestNotificationAgreementOptions {
  options: { templateCode: string };
  onEvent: (result: { type: NotificationAgreementResult }) => void;
  onError: (error: unknown) => void | Promise<void>;
}

/**
 * @public
 * @category 알림
 * @name requestNotificationAgreement
 * @description 알림 동의 화면을 요청하고, 동의 결과를 콜백으로 전달해요.
 * @param {RequestNotificationAgreementOptions} params - 템플릿 코드와 콜백이에요.
 * @returns {() => void} 앱브릿지 콜백을 해제하는 함수예요.
 *
 * @example
 * ```ts
 * import { requestNotificationAgreement } from '@apps-in-toss/framework';
 *
 * const cleanup = requestNotificationAgreement({
 *   options: { templateCode: 'test-template-code' },
 *   onEvent: (result) => {
 *     console.log(result);
 *     cleanup();
 *   },
 *   onError: (error) => {
 *     console.error(error);
 *     cleanup();
 *   },
 * });
 * ```
 */
export function requestNotificationAgreement(params: RequestNotificationAgreementOptions): () => void {
  if (!isRequestNotificationAgreementSupported()) {
    return noop;
  }

  const { options, onEvent, onError } = params;

  return INTERNAL__appBridgeHandler.invokeAppBridgeMethod('requestNotificationAgreement', options, {
    onSuccess: (result: { agreementResult: NotificationAgreementResult }) => {
      onEvent({ type: result.agreementResult });
    },
    onError: (error: unknown) => {
      onError(error);
    },
  });
}
