// @base-ui/react/esm/otp-field/input/OTPFieldInputDataAttributes.js
export let OTPFieldInputDataAttributes = /*#__PURE__*/function (OTPFieldInputDataAttributes) {
  /**
   * Present when all slots are filled.
   */
  OTPFieldInputDataAttributes["complete"] = "data-complete";
  /**
   * Present when the input contains a character.
   */
  OTPFieldInputDataAttributes["filled"] = "data-filled";
  /**
   * Present when the OTP field is disabled.
   */
  OTPFieldInputDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the OTP field is readonly.
   */
  OTPFieldInputDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the OTP field is required.
   */
  OTPFieldInputDataAttributes["required"] = "data-required";
  /**
   * Present when the OTP field is in a valid state (when wrapped in Field.Root).
   */
  OTPFieldInputDataAttributes["valid"] = "data-valid";
  /**
   * Present when the OTP field is in an invalid state (when wrapped in Field.Root).
   */
  OTPFieldInputDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the OTP field has been touched (when wrapped in Field.Root).
   */
  OTPFieldInputDataAttributes["touched"] = "data-touched";
  /**
   * Present when the OTP field's value has changed (when wrapped in Field.Root).
   */
  OTPFieldInputDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when any OTP field input is focused.
   */
  OTPFieldInputDataAttributes["focused"] = "data-focused";
  return OTPFieldInputDataAttributes;
}({});
